// @likashefqet/react-native-image-zoom/lib/module/components/ImageZoom.js
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
import React, { forwardRef } from 'react';
import { StyleSheet } from 'react-native';
import { GestureDetector } from 'react-native-gesture-handler';
import Animated from 'react-native-reanimated';
import { useZoomable } from '../hooks/useZoomable';
const styles = StyleSheet.create({
  image: {
    flex: 1
  }
});
const Zoomable = (_ref, ref) => {
  let {
    uri = '',
    minScale,
    maxScale,
    scale,
    doubleTapScale,
    maxPanPointers,
    isPanEnabled,
    isPinchEnabled,
    isSingleTapEnabled,
    isDoubleTapEnabled,
    onInteractionStart,
    onInteractionEnd,
    onPinchStart,
    onPinchEnd,
    onPanStart,
    onPanEnd,
    onSingleTap,
    onDoubleTap,
    onProgrammaticZoom,
    onResetAnimationEnd,
    onLayout,
    style = {},
    ...props
  } = _ref;
  const {
    animatedStyle,
    gestures,
    onZoomableLayout
  } = useZoomable({
    minScale,
    maxScale,
    scale,
    doubleTapScale,
    maxPanPointers,
    isPanEnabled,
    isPinchEnabled,
    isSingleTapEnabled,
    isDoubleTapEnabled,
    onInteractionStart,
    onInteractionEnd,
    onPinchStart,
    onPinchEnd,
    onPanStart,
    onPanEnd,
    onSingleTap,
    onDoubleTap,
    onProgrammaticZoom,
    onResetAnimationEnd,
    onLayout,
    ref
  });
  return /*#__PURE__*/React.createElement(GestureDetector, {
    gesture: gestures
  }, /*#__PURE__*/React.createElement(Animated.Image, _extends({
    style: [styles.image, style, animatedStyle],
    source: {
      uri
    },
    resizeMode: "contain",
    onLayout: onZoomableLayout
  }, props)));
};
export default /*#__PURE__*/forwardRef(Zoomable);
